eth-contracts/contracts/service/DelegateManager.sol

```diff
@@ -0,0 +1,556 @@
+        require(
+            undelegateRequests[delegator].lockupExpiryBlock <= block.number, "Lockup must be expired");
+
+        // Confirm no pending claim for this service provider
+        require(
+            claimPending(undelegateRequests[delegator].serviceProvider) == false,
+            "Undelegate not permitted for SP pending claim"
+        );
+
+        address serviceProvider = undelegateRequests[delegator].serviceProvider;
+        uint unstakeAmount = undelegateRequests[delegator].amount;
+
+        bool exists = delegatorExistsForSP(delegator, serviceProvider);
+        require(exists, "Delegator must be staked for SP");
+
+        Staking stakingContract = Staking(
+            registry.getContract(stakingProxyOwnerKey)
+        );
+
+        // Stake on behalf of target service provider
+        stakingContract.undelegateStakeFor(
+            serviceProvider,
+            delegator,
+            unstakeAmount,
+            empty
+        );
+
+        // Update amount staked from this delegator to targeted service provider
+        delegateInfo[delegator][serviceProvider] -= unstakeAmount;
+
+        // Update total delegated stake
+        delegatorStakeTotal[delegator] -= unstakeAmount;
+
+        // Update total delegated for SP
+        spDelegateInfo[serviceProvider].totalDelegatedStake -= unstakeAmount;
+
+        // Remove from delegators list if no delegated stake remaining
+        if (delegateInfo[delegator][serviceProvider] == 0) {
+            bool foundDelegator;
+            uint delegatorIndex;
+            for (uint i = 0; i < spDelegateInfo[serviceProvider].delegators.length; i++) {
+                if (spDelegateInfo[serviceProvider].delegators[i] == delegator) {
+                    foundDelegator = true;
+                    delegatorIndex = i;
+                }
+            }
+
+            if (foundDelegator) {
+                // Overwrite and shrink delegators list
+                uint lastIndex = spDelegateInfo[serviceProvider].delegators.length - 1;
+                spDelegateInfo[serviceProvider].delegators[delegatorIndex] = spDelegateInfo[serviceProvider].delegators[lastIndex];
+                spDelegateInfo[serviceProvider].delegators.length--;
+            }
+        }
+
+        // Update total locked for this service provider
+        spDelegateInfo[serviceProvider].totalLockedUpStake -= unstakeAmount;
+
+        // Reset lockup information
+        undelegateRequests[delegator] = UndelegateStakeRequest({
+            lockupExpiryBlock: 0,
+            amount: 0,
+            serviceProvider: address(0)
+        });
+
+        // Validate balance
+        ServiceProviderFactory(
+            registry.getContract(serviceProviderFactoryKey)
+        ).validateAccountStakeBalance(serviceProvider);
+
+        // Return new total
+        return delegateInfo[delegator][serviceProvider];
+    }
+
+    /*
+      TODO: See if its worth splitting processClaim into a separate tx?
```

Comment:
Independent - validate w/real world gas price, block gas limit is already 8,000,000 which is < mainnet so no issues there